src/axom/core/ArrayBase.hpp

```diff
@@ -533,53 +534,364 @@ struct all_types_are_integral
   static constexpr bool value = all_types_are_integral_impl<Args...>::value;
 };
 
+template <bool has_default_ctor>
+struct OpInitBase
+{
+  template <typename T, typename ExecSpace>
+  static void init(T* data, IndexType begin, IndexType end);
+};
+
+template <bool fill_on_host>
+struct OpFillBase
+{
+  template <typename T, typename ExecSpace>
+  static void fill(T* array, IndexType n, const T& value);
+};
+
+template <bool destroy_on_host>
+struct OpDestroyBase
+{
+  template <typename T, typename ExecSpace>
+  static void destroy(T* array, IndexType begin, IndexType end);
+};
+
+template <bool device>
+struct OpMemmoveBase
+{
+  template <typename T>
+  static void move(T* array,
+                   IndexType src_begin,
+                   IndexType src_end,
+                   IndexType dst,
+                   int allocId);
+};
+
 /*!
  * \brief Default-initializes the "new" segment of an array
  *
  * \param [inout] data The data to initialize
- * \param [in] pos The beginning of the subset of \a data that should be initialized
- * \param [in] len The length of the subset of \a data that is to be initialized
- * \param [in] alloc_id The allocator ID with which \a data was allocated
- * 
- * The final parameter is intended to be used via tag dispatch with std::is_default_constructible
+ * \param [in] begin The beginning of the subset of \a data that should be initialized
+ * \param [in] end The end index (exclusive) of the subset of \a data that is to be initialized
  */
-template <typename T>
-void initializeInPlace(T* data,
-                       const IndexType pos,
-                       const IndexType len,
-                       const int alloc_id,
-                       std::true_type)
+template <>
+template <typename T, typename ExecSpace>
+void OpInitBase<true>::init(T* data, IndexType begin, IndexType end)
 {
-#if defined(__CUDACC__) && defined(AXOM_USE_UMPIRE) && \
-  defined(UMPIRE_ENABLE_DEVICE)
-  if(alloc_id == getAllocatorID<MemorySpace::Device>())
+#if defined(__CUDACC__) && defined(AXOM_USE_UMPIRE)
+  if(axom::execution_space<ExecSpace>::onDevice())
   {
+    IndexType len = end - begin;
     // If we instantiated a fill kernel here it would require
     // that T's default ctor is device-annotated which is too
     // strict of a requirement, so we copy a buffer instead.
     T* tmp_buffer = new T[len]();
-    axom::copy(data + pos, tmp_buffer, len * sizeof(T));
+    axom::copy(data + begin, tmp_buffer, len * sizeof(T));
     delete[] tmp_buffer;
     return;
   }
-#else
-  AXOM_UNUSED_VAR(alloc_id);
 #endif
-  for(int ielem = 0; ielem < len; ielem++)
+  for(int ielem = begin; ielem < end; ielem++)
```

Comment:
Although it doesn't matter here since ielem is an int, I believe that stylistically we prefer pre-fix '++' for consistency across the code.